### ISSUE TITLE
[Megatron][MoE] grad_norm explosion under context parallel (CP>1): loss_func must return 3-tuple when calculate_per_token_loss=True

### ISSUE THREAD

**Issue by insightLLM** [verl-project/verl#6609]
## Summary

When training a **MoE** model with the **Megatron** backend under **context parallelism (`CP > 1`)** together with `use_remove_padding=True`, `actor/grad_norm` explodes to the **thousands** and validation scores steadily degrade. With `CP=1` the exact same config is healthy (`grad_norm` ~0.3).

Root cause: verl's actor `loss_func` always returns a **2-tuple** `(loss, metrics)`. But when Megatron runs with `calculate_per_token_loss=True` (which some models, e.g. Qwen3-VL via megatron-bridge, **force** under CP), the loss function is expected to return a **3-tuple** `(loss_sum, num_tokens, metrics)`. Because verl never provides `num_tokens`, the per-token normalization in `finalize_model_grads` is skipped, and the **MoE aux/z-loss gradients are over-scaled by ~`num_tokens` (tens of thousands)**, which dominates the policy gradient.

## Environment

- verl: `0.8.0.dev`
- Megatron-LM: https://github.com/insightLLM/Megatron-LM/tree/mtp-local-changes

- megatron-bridge : 0.4.1
- Model: Qwen3.5-style MoE (MoE + Gated Delta Net), `~35B-A3B`
- Parallelism: `TP/PP/EP/...` with `CP=2`, `use_remove_padding=True`
- Symptom only appears at `CP > 1`.

## Symptom

```
# CP=2 (broken)
actor/grad_norm: 3168.13...
val-core/.../acc/mean@1: decreasing over steps

# CP=1 (healthy), or CP=2 with moe_aux_loss_coeff=0 & moe_z_loss_coeff=0
actor/grad_norm: ~0.33
```

Setting `moe_aux_loss_coeff=0` and `moe_z_loss_coeff=0` at `CP=2` makes `grad_norm` collapse from ~3000 back to ~0.33 — i.e. the explosion comes from the **MoE auxiliary / z losses**, not the policy loss.

## Root cause

Three pieces interact:

1. **Some models force `calculate_per_token_loss=True` under CP.** For these models, disabling the flag is not an option (hard assert: `"... only supports context parallelism with calculate_per_token_loss enabled"`). With `CP=1` the flag is *not* forced, which is exactly why the bug is CP-only.

2. **The MoE router pre-scales aux/z loss by the local token count** when `calculate_per_token_loss=True` (`megatron/core/transformer/moe/router.py`):

```python
if self.calculate_per_token_loss:
    num_tokens = valid_token_count if valid_token_count is not None else activation.shape[0]
    activation = MoEAuxLossAutoScaler.apply(activation, aux_loss * num_tokens)
```

   This multiplication is intentional: it assumes the final gradient will later be divided by the global token count.

3. **That division only happens if `loss_func` returns a 3-tuple.** In `megatron/core/pipeline_parallel/schedules.py`:

```python
outputs = loss_func(output_tensor)
if len(outputs) == 3:
    output_tensor, num_tokens, loss_reduced = outputs
    # total_num_tokens accumulates here -> used by finalize_model_grads to divide all grads
    ...
else:
    assert len(outputs) == 2          # <-- verl's path
    output_tensor, loss_reduced = outputs
    output_tensor *= cp_group_size
    output_tensor /= num_microbatches
```

   With `calculate_per_token_loss=True`, DDP sets `gradient_scaling_factor = 1.0` (it asserts `average_in_collective=False`), so the *only* place the `* num_tokens` pre-scaling of the aux/z loss gets cancelled is the `1/total_num_tokens` division inside `finalize_model_grads`. Because verl returns a 2-tuple, `total_num_tokens` is never accumulated and that division never runs. The aux/z gradient therefore stays multiplied by ~`num_tokens` (tens of thousands), exploding `grad_norm` and swamping the policy gradient → eval degradation.

This is invisible at `CP=1` only because the flag is not forced there.

## Fix

Make verl's Megatron actor `loss_func` return the 3-tuple `(loss_sum, num_tokens, metrics)` **when `calculate_per_token_loss` is enabled**, so Megatron's per-token-loss pipeline normalizes policy + aux + z gradients by the global token count. The legacy 2-tuple path is kept untouched for the common `calculate_per_token_loss=False` case.

In `verl/workers/actor/megatron_actor.py`, inside `loss_func`, replace the final `return policy_loss, [metrics, ret_entropy]` with:

```python
if not forward_only and getattr(self.tf_config, "calculate_per_token_loss", False):
    cp_size = mpu.get_context_parallel_world_size()
    # response_mask is the full (CP-gathered) response, so its sum is identical on every
    # CP rank. Divide by cp_size so the dp_cp all-reduce inside finalize_model_grads
    # recovers the true global token count instead of double-counting CP ranks.
    full_num_tokens = response_mask.sum()
    # int32 to match Megatron's `total_num_tokens` accumulator (avoids in-place
    # int32 += int64/float dtype-cast errors in the pipeline schedule).
    num_tokens = (full_num_tokens // cp_size).clamp_min(1).detach().to(torch.int32)
    # Scale the (seq-mean-token-mean) loss to token-sum scale so finalize's
    # 1/total_num_tokens recovers the intended per-token-mean gradient and cancels the
    # router's aux/z `* num_local_tokens` pre-scaling.
    loss_sum = policy_loss * full_num_tokens
    return loss_sum, num_tokens, [metrics, ret_entropy]

return policy_loss, [metrics, ret_entropy]
```

Key points:
- `num_tokens` is divided by `cp_size` because `response_mask` is the CP-gathered (full) response and the `dp_cp` all-reduce inside `finalize_model_grads` would otherwise count each CP rank.
- `num_tokens` is cast to `int32` to match Megatron's `total_num_tokens` accumulator dtype.
- `loss_sum = policy_loss * full_num_tokens` puts the loss on a token-sum scale so that finalize's `1/total_num_tokens` recovers the intended per-token-mean gradient and cancels the router's `* num_local_tokens` pre-scaling.

## Result

After the fix, at `CP=2` `actor/grad_norm` returns to ~0.33 and validation scores recover/improve — matching `CP=1` behavior.

## Proposed PR

Happy to send a PR (I have the change running). The change is gated behind `calculate_per_token_loss` so it is a no-op for existing non-CP / non-per-token-loss setups.

https://github.com/insightLLM/verl/tree/qwen3_5-gdn-thd-mm-agentloop-fixes

**Comment by insightLLM** [comment]
The yellow line represents CP=2,  compared with the gray line represents CP=1. 

run the train:
https://github.com/insightLLM/verl/blob/qwen3_5-gdn-thd-mm-agentloop-fixes/script/run_qwen3_5-35b-megatron.sh

<img width="1000" height="400" alt="Image" src="https://github.com/user-attachments/assets/a15e7480-8579-45e3-8773-eed30f8eb044" />

<img width="400" height="400" alt="Image" src="https://github.com/user-attachments/assets/008a0b1e-b547-48e1-8467-d8ec2db65102" />

<img width="400" height="400" alt="Image" src="https://github.com/user-attachments/assets/d5b9cab8-d410-4d16-9549-04f5249282d9" />

It can be seen that the two are very close on the final evaluation set.

<img width="400" height="400" alt="Image" src="https://github.com/user-attachments/assets/6ace4cac-fb39-4516-a9e3-4ba588991a84" />

**Comment by wuxibin89** [comment]
> Some models force calculate_per_token_loss=True under CP. For these models, disabling the flag is not an option (hard assert: "... only supports context parallelism with calculate_per_token_loss enabled"). With CP=1 the flag is not forced, which is exactly why the bug is CP-only.

@HollowMan6 Why megatron-bridge enforce calculate_per_token_loss=True under CP?

**Comment by HollowMan6** [comment]
It looks like Megatron-Bridge only hard code it for https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/9f69b72de8a922caa93ebf2f33e13f0c6399753e/src/megatron/bridge/models/qwen_vl/modelling_qwen3_vl/model.py#L202-L205, probably would need to check with the original Qwen3 VL contributor to get the reason.

btw, would you mind let us know if #6836 works for you @insightLLM ?

**Comment by yaoyu-33** [comment]
This is essentially a Megatron-Core loss-normalization requirement for CP training, especially for masked / packed finetuning workloads.

With context parallelism, different CP ranks may own different numbers of valid/loss tokens. The legacy 2-tuple loss path cannot correctly normalize by the global non-padded token count in those cases. `calculate_per_token_loss=True` switches Megatron-Core to the per-token normalization contract, where the loss function returns `(loss_sum, num_tokens, metrics)` and `finalize_model_grads` divides gradients by the accumulated global token count.

For reference, NeMo-RL hit a similar CP/MCore normalization requirement and now sets `calculate_per_token_loss=True` in the Megatron setup path, with `average_in_collective=False` for DDP:

https://github.com/NVIDIA-NeMo/RL/blob/main/nemo_rl/models/megatron/setup.py#L879-L885
https://github.com/NVIDIA-NeMo/RL/blob/main/nemo_rl/models/megatron/setup.py#L975-L978

This was part of the Qwen3.5 CP support work:
https://github.com/NVIDIA-NeMo/RL/pull/2312

**Comment by EricMarcus-ai** [comment]
> It looks like Megatron-Bridge only hard code it for https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/9f69b72de8a922caa93ebf2f33e13f0c6399753e/src/megatron/bridge/models/qwen_vl/modelling_qwen3_vl/model.py#L202-L205, probably would need to check with the original Qwen3 VL contributor to get the reason.
> 
> btw, would you mind let us know if [#6836](https://github.com/verl-project/verl/pull/6836) works for you [@insightLLM](https://github.com/insightLLM) ?

If we get no reply, I think we could merge @HollowMan6. The issue described is precisely what I experienced, which caused this PR. I already made the fix several weeks ago; I only got around to PR'ing it now, and that is when I found this issue. Let me know if any changes needed.

**Comment by HollowMan6** [comment]
I created another way of workaround which instead scale the loss by dividing dp_cp_world_size https://github.com/verl-project/verl/pull/6951/changes/785b8632858206814277c030a147c95aff140747. Please let me know if this works for you or not @EricMarcus-ai @insightLLM

**Comment by EricMarcus-ai** [comment]
@HollowMan6, I think that fixes the policy-loss scaling, not the MoE aux/z blowup this issue is about. Your `not has_auxiliary_loss` guard skips exactly that case, and I think it can't be extended to cover it: the aux/z gradient comes from `MoEAuxLossAutoScaler` (bypasses `scaled_loss`) and its blowup factor is the variable per-microbatch `num_tokens`, which no constant `1/(dp*cp)` can cancel. That needs the token-count division, i.e. the 3-tuple `(loss_sum, num_tokens, output)` contract so `finalize_model_grads` handles it, which is what #6836 does and what your comment already defers to.
